rafthttp/peer.go

```diff
@@ -15,158 +15,196 @@
+			}
+		case mm := <-p.recvc:
+			p.msgAppDecoder.updateTerm(mm.Term)
+			if err := p.r.Process(context.TODO(), mm); err != nil {
+				log.Printf("peer: process raft message error: %v", err)
+			}
+		case u := <-p.updatec:
+			p.msgAppDialer.update(u)
+			p.dialer.update(u)
+			p.pipeline.update(u)
+		case <-p.pausec:
+			paused = true
+		case <-p.resumec:
+			paused = false
+		case <-p.stopc:
+			p.msgAppReader.stop()
+			p.msgAppWriter.stop()
+			p.reader.stop()
+			p.writer.stop()
+			p.pipeline.stop()
+			close(p.done)
+			return
+		}
 	}
-	p.u = u
-	p.pipeline.update(u)
 }
 
-// Send sends the data to the remote node. It is always non-blocking.
-// It may be fail to send data if it returns nil error.
-// TODO (xiangli): reasonable retry logic
-func (p *peer) Send(m raftpb.Message) error {
-	p.Lock()
-	defer p.Unlock()
-	if p.stopped {
-		return errors.New("peer: stopped")
-	}
-	if p.paused {
-		return nil
-	}
-
-	// move all the stream related stuff into stream
-	p.stream.invalidate(m.Term)
-	if shouldInitStream(m) && !p.stream.isOpen() {
-		u := p.u
-		// todo: steam open should not block.
-		p.stream.open(types.ID(m.From), p.id, p.cid, m.Term, p.tr, u, p.r)
-		p.batcher.Reset(time.Now())
+func (p *peer) Send(m raftpb.Message) {
+	select {
+	case p.sendc <- m:
+	case <-p.done:
+		log.Panicf("peer: unexpected stopped")
 	}
+}
 
-	var err error
-	switch {
-	case isProposal(m):
-		p.propBatcher.Batch(m)
-	case canBatch(m) && p.stream.isOpen():
-		if !p.batcher.ShouldBatch(time.Now()) {
-			err = p.pipeline.send(m)
-		}
-	case canUseStream(m):
-		if ok := p.stream.write(m); !ok {
-			err = p.pipeline.send(m)
-		}
+func (p *peer) recv(m raftpb.Message) error {
+	select {
+	case p.recvc <- m:
+		return nil
+	case <-p.done:
+		return fmt.Errorf("peer is stopped")
 	default:
-		err = p.pipeline.send(m)
+		log.Printf("peer: dropping %s because maximal number %d of receiver buffer entries from %s has been reached",
+			m.Type, recvBufSize, p.to)
+		return fmt.Errorf("recv buffer is full")
 	}
-	// send out batched MsgProp if needed
-	// TODO: it is triggered by all outcoming send now, and it needs
-	// more clear solution. Either use separate goroutine to trigger it
-	// or use streaming.
-	if !p.propBatcher.IsEmpty() {
-		t := time.Now()
-		if !p.propBatcher.ShouldBatch(t) {
-			p.pipeline.send(p.propBatcher.Message)
-			p.propBatcher.Reset(t)
-		}
-	}
-	return err
 }
 
-// Stop performs any necessary finalization and terminates the peer
-// elegantly.
-func (p *peer) Stop() {
-	p.Lock()
-	defer p.Unlock()
-	p.pipeline.stop()
-	p.stream.stop()
-	p.stopped = true
+func (p *peer) Update(u string) {
+	select {
+	case p.updatec <- u:
+	case <-p.done:
+		log.Panicf("peer: unexpected stopped")
+	}
 }
 
-// attachStream attaches a streamSever to the peer.
-func (p *peer) attachStream(sw *streamWriter) error {
-	p.Lock()
-	defer p.Unlock()
-	if p.stopped {
-		return errors.New("peer: stopped")
+func (p *peer) attachOutgoingConn(conn outgoingConn) {
```

Comment:
the run loop is designed as non-blocking. and the msgAppWriter.attchOutgoingConn and writer.attachoOutgoigConn are blocking function. so I don't let it go into the run loop.